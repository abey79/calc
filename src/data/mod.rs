pub mod ast;
pub mod identified;
pub mod span;
pub mod token;
