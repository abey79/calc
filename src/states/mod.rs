//! This module models states (as in "state machine") between the compilation steps.
//!
//! The data associated with each state is factored in "context" structure for reusability (e.g. the
//! [`TokenStream`] context ("state data") is used in both the [`TokenizedState`] and the
//! [`ParsedState`]).

pub mod input_state;
pub mod parsed_state;
pub mod tokenized_state;

pub use input_state::*;
pub use parsed_state::*;
pub use tokenized_state::*;
