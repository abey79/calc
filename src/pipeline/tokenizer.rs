//! The tokenizer stage.
//!
//! Transforms a `RawInput` into a `TokenizedInput`.

use crate::context::token_stream::TokenStream;
use crate::data::span::{Loc, Span};
use crate::data::token::{Token, TokenKind};
use crate::errors::{Spanned, SyntaxError, TokenizerError};
use crate::states::{InputState, TokenizedState};

type Result<T> = std::result::Result<T, TokenizerError>;

pub(crate) fn tokenize(input: InputState) -> Result<TokenizedState> {
    let mut tokenizer = Tokenizer::new(input);
    tokenizer.run()?;
    Ok(TokenizedState {
        source: tokenizer.input.source,
        token_stream: tokenizer.token_stream,
    })
}

/// Implement the tokenizer stage.
struct Tokenizer {
    input: InputState,

    /// current position in the input, updated by [`next()`]
    pos: usize,

    /// current location in the input, updated by [`next()`]
    loc: Loc,

    /// starting location of the current token, updated by the lexer loop
    start_loc: Loc,

    /// current stream of token
    token_stream: TokenStream,
}

impl Tokenizer {
    fn new(input: InputState) -> Self {
        Self {
            input,
            pos: 0,
            loc: Loc::default(),
            start_loc: Loc::default(),
            token_stream: TokenStream::default(),
        }
    }

    fn source(&self) -> &str {
        self.input.as_ref()
    }

    /// Push a token into the token context.
    fn push(&mut self, kind: TokenKind) {
        let span = Span::new(self.start_loc, self.loc);
        let token = Token::new(kind, span);
        self.token_stream.push_token(token);
    }

    /// Return the next character in the input stream and update the current location.
    ///
    /// Returns `None` if the end of the input is reached.
    fn next(&mut self) -> Option<char> {
        let c = self.source().chars().nth(self.pos);
        if let Some(c) = c {
            self.pos += 1;
            if c == '\n' {
                self.loc.line += 1;
                self.loc.col = 0;
            } else {
                self.loc.col += 1;
            }
        }

        c
    }

    /// Return the next character in the input stream without updating the current location.
    ///
    /// Returns `None` if the end of the input is reached.
    fn peek(&self) -> Option<char> {
        self.source().chars().nth(self.pos)
    }

    /// Return the next character in the input stream if it matches `c` and update the current
    /// location.
    fn accept(&mut self, c: char) -> bool {
        if self.peek() == Some(c) {
            self.next();
            true
        } else {
            false
        }
    }

    fn err<T>(&self, err: SyntaxError) -> Result<T> {
        let span = Span::new(self.loc, self.loc);
        let new_err = TokenizerError::SyntaxError(err, span.to_error(&self.input.source));

        Err(new_err)
    }

    fn run(&mut self) -> Result<()> {
        while let Some(c) = self.next() {
            self.start_loc = self.loc;

            match c {
                // whitespace
                c if c.is_whitespace() => continue,
                // integer/float
                c if c.is_ascii_digit() => {
                    let mut num = c.to_string();
                    while let Some(c) = self.peek() {
                        if c.is_ascii_digit() {
                            num.push(c);
                            self.next();
                        } else {
                            break;
                        }
                    }
                    if let Some(c) = self.peek() {
                        if c == '.' {
                            num.push(c);
                            self.next();
                            while let Some(c) = self.peek() {
                                if c.is_ascii_digit() {
                                    num.push(c);
                                    self.next();
                                } else {
                                    break;
                                }
                            }
                            self.push(TokenKind::Float(num.parse().unwrap()));
                        } else {
                            self.push(TokenKind::Int(num.parse().unwrap()));
                        }
                    } else {
                        self.push(TokenKind::Int(num.parse().unwrap()));
                    }
                }
                // names/keywords
                c if c.is_ascii_alphabetic() || c == '_' => {
                    let mut name = c.to_string();
                    while let Some(c) = self.peek() {
                        if c.is_ascii_alphanumeric() || c == '_' {
                            name.push(c);
                            self.next();
                        } else {
                            break;
                        }
                    }
                    match name.as_str() {
                        // keywords
                        "print" => self.push(TokenKind::Print),
                        _ => self.push(TokenKind::Name(name)),
                    }
                }
                // misc
                ';' => self.push(TokenKind::Semi),
                '(' => self.push(TokenKind::LParen),
                ')' => self.push(TokenKind::RParen),
                ',' => self.push(TokenKind::Comma),
                '=' => self.push(TokenKind::Assign),
                '+' => self.push(TokenKind::Plus),
                '-' => self.push(TokenKind::Minus),
                '*' => self.push(TokenKind::Star),
                '/' => {
                    if self.accept('/') {
                        while let Some(c) = self.next() {
                            if c == '\n' {
                                break;
                            }
                        }
                    } else if self.accept('*') {
                        while let Some(c) = self.next() {
                            if c == '*' && self.peek() == Some('/') {
                                self.next();
                                break;
                            }
                        }
                    } else {
                        self.push(TokenKind::Slash);
                    }
                }

                c => return self.err(SyntaxError::UnexpectedCharacter(c)),
            }
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tokenize() {
        let input = InputState::from("a = (1.3 + 3.2) * 45.1; b = a * 3.2; print 1 + 2 * 3;");
        let tokenized = tokenize(input).unwrap();

        insta::assert_debug_snapshot!(tokenized.token_stream);
    }
}
