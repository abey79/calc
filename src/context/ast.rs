use crate::data::ast::Stmt;
use crate::data::meta::Meta;
use crate::data::span::Span;
use crate::data::token_span::TokSpan;
use crate::pipeline;
use std::fmt;
use std::fmt::{Debug, Display, Write};

#[derive(Debug, Default)]
pub struct Ast<M: Debug + Display> {
    stmts: Vec<Stmt<M>>,
}

impl<M: Debug + Display> Ast<M> {
    pub fn new() -> Self {
        Self { stmts: Vec::new() }
    }

    pub fn stmts(&self) -> &[Stmt<M>] {
        &self.stmts
    }

    pub fn stmts_mut(&mut self) -> &mut Vec<Stmt<M>> {
        &mut self.stmts
    }

    pub fn push_stmt(&mut self, stmt: Stmt<M>) {
        self.stmts.push(stmt);
    }

    pub fn format<W: fmt::Write>(&self, w: &mut W) -> Result<String, fmt::Error> {
        pipeline::formatter::format(self, w)
    }
}

impl Ast<TokSpan> {
    pub fn dump<W: Write>(&self, w: &mut W) -> std::fmt::Result {
        for stmt in self.stmts() {
            writeln!(w, "\n{:#?}", stmt)?;
        }

        Ok(())
    }
}

// Provide convenient methods for AST nodes parametrized by a token span.
impl<K> Meta<K, TokSpan> {
    pub fn tok_span(&self) -> TokSpan {
        self.meta.clone()
    }

    pub fn span(&self) -> Span {
        self.meta.span()
    }
}
