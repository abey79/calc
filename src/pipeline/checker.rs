use crate::context::checked_ast::{CheckedAst, CheckedExpr, CheckedStmt, Type, TypeInfo};
use crate::data::ast::{BinOp, Expr, ExprKind, Stmt, StmtKind, UnaryOp, VarName};
use crate::data::meta::Meta;
use crate::data::token_span::TokSpan;
use crate::errors::error_message::Spanned;
use crate::errors::{CheckerError, SyntaxError, TypeError};
use crate::states::{CheckedState, ParsedState};
use std::collections::HashMap;

type Result<T> = std::result::Result<T, CheckerError>;

pub(crate) fn check(input: ParsedState) -> Result<CheckedState> {
    let mut checker = Checker::new(&input);
    let checked_ast = checker.run()?;
    Ok(CheckedState {
        source: input.source,
        token_stream: input.token_stream,
        raw_ast: input.raw_ast,
        ast: checked_ast,
    })
}

struct Checker<'a> {
    input: &'a ParsedState,

    // state
    vars: HashMap<String, Type>,
}

impl<'a> Checker<'a> {
    fn new(input: &'a ParsedState) -> Self {
        Self {
            input,
            vars: HashMap::new(),
        }
    }

    fn run(&mut self) -> Result<CheckedAst> {
        let mut checked_ast = CheckedAst::new();
        for stmt in self.input.raw_ast.stmts() {
            checked_ast.push_stmt(self.check_stmt(stmt)?);
        }

        Ok(checked_ast)
    }

    fn check_stmt(&mut self, stmt: &Stmt<TokSpan>) -> Result<CheckedStmt> {
        match &stmt.kind {
            StmtKind::Assign { name, value } => {
                let checked_value = self.check_expr(value)?;
                let type_ = checked_value.meta.type_;
                self.vars.insert(name.kind.clone(), type_);
                Ok(Stmt::assign(
                    VarName::new(&name.kind, TypeInfo::new(type_, name.tok_span())),
                    checked_value,
                    TypeInfo::new(Type::Stmt, stmt.tok_span()),
                ))
            }
            StmtKind::Print { expr } => {
                let checked_expr = self.check_expr(expr)?;
                Ok(Stmt::print(
                    checked_expr,
                    TypeInfo::new(Type::Stmt, stmt.tok_span()),
                ))
            }
            StmtKind::Expr { expr } => {
                let checked_expr = self.check_expr(expr)?;
                Ok(Stmt::expr(
                    checked_expr,
                    TypeInfo::new(Type::Stmt, stmt.tok_span()),
                ))
            }
        }
    }

    fn check_expr(&mut self, expr: &Expr<TokSpan>) -> Result<CheckedExpr> {
        match &expr.kind {
            ExprKind::Variable(name) => {
                if let Some(type_) = self.vars.get(&name.kind) {
                    Ok(Expr::variable(
                        VarName::new(&name.kind, TypeInfo::new(*type_, name.tok_span())),
                        TypeInfo::new(*type_, expr.tok_span()),
                    ))
                } else {
                    Err(self.syntax_err(SyntaxError::UnknownVariable(name.kind.clone()), expr))
                }
            }
            ExprKind::BinOp { op, left, right } => {
                let checked_left = self.check_expr(left)?;
                let checked_right = self.check_expr(right)?;
                let left_type = checked_left.meta.type_;
                let right_type = checked_right.meta.type_;

                if left_type != right_type {
                    return Err(self.type_err(
                        TypeError::MismatchedTypesForBinaryOp(left_type, right_type),
                        op,
                    ));
                }

                Ok(Expr::bin_op(
                    BinOp::new(op.kind, TypeInfo::new(left_type, op.tok_span())),
                    checked_left,
                    checked_right,
                    TypeInfo::new(left_type, expr.tok_span()),
                ))
            }
            ExprKind::UnaryOp { op, operand } => {
                let checked_expr = self.check_expr(operand)?;
                let type_ = checked_expr.meta.type_;

                if !matches!(type_, Type::Integer | Type::Float) {
                    return Err(self.type_err(TypeError::InvalidTypeForUnaryOp(type_), expr));
                }

                Ok(Expr::unary_op(
                    UnaryOp::new(op.kind, TypeInfo::new(type_, op.tok_span())),
                    checked_expr,
                    TypeInfo::new(type_, expr.tok_span()),
                ))
            }
            ExprKind::Tuple(..) => todo!(),
            ExprKind::Integer(i) => Ok(Expr::integer(
                *i,
                TypeInfo::new(Type::Integer, expr.tok_span()),
            )),
            ExprKind::Float(fl) => Ok(Expr::float(
                *fl,
                TypeInfo::new(Type::Float, expr.tok_span()),
            )),
        }
    }

    fn type_err<K>(&self, err: TypeError, node: &Meta<K, TokSpan>) -> CheckerError {
        CheckerError::TypeError(err, node.to_error(&self.input.source))
    }

    fn syntax_err<K>(&self, err: SyntaxError, node: &Meta<K, TokSpan>) -> CheckerError {
        CheckerError::SyntaxError(err, node.to_error(&self.input.source))
    }
}

#[cfg(test)]
mod tests {
    use crate::pipeline::checker::check;
    use crate::pipeline::parser::parse;
    use crate::pipeline::tokenizer::tokenize;
    use crate::states::InputState;

    #[test]
    fn test_checker() {
        let input = InputState::from("a = (1.3 + 3.2) * 45.1; b = a * 3.2; print 1 + 2 * 3;");
        let tokenized = tokenize(input).unwrap();
        let parsed = parse(tokenized).unwrap();
        let checked = check(parsed).unwrap();

        insta::assert_debug_snapshot!(checked.ast);
    }
}
