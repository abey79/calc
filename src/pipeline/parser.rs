use crate::data::ast::{BinOp, Expr, Stmt, UnaryOp, VarName};
use crate::data::token::{Token, TokenId, TokenKind};
use crate::errors::{ParserError, SyntaxError};
use crate::states::{AstContext, ParsedInput, TokenizedInput};
use std::rc::Rc;

type Result<T> = std::result::Result<T, ParserError>;

pub(crate) fn parse(input: TokenizedInput) -> Result<ParsedInput> {
    let mut parser = Parser::new(input);
    parser.run()?;
    Ok(ParsedInput {
        text_ctx: parser.input.text_ctx,
        token_ctx: parser.input.token_ctx,
        ast_ctx: parser.ast_ctx,
    })
}

/// Returns the next token if it matches and advances the parser, or returns an error.
macro_rules! expect {
    ($self:ident, $kind:pat) => {{
        let token = $self
            .tokens()
            .get($self.pos)
            .ok_or_else(|| $self.end_of_file_err())?;

        if matches!(token.kind, $kind) {
            Ok($self.next().unwrap())
        } else {
            Err(ParserError::SyntaxError(
                SyntaxError::UnexpectedToken(token.kind.clone()),
                $self
                    .input
                    .text_ctx
                    .error_context($self.input.token_ctx.span_from_id(token.id)),
            ))
        }
    }};
}

/// Returns an `Option<Token>` if the token matches, in which case the parser advances
macro_rules! accept {
    ($self:ident, $kind:pat) => {{
        $self.pos += 1; // must be moved before the get to avoid double-borrowing
        let token = $self.tokens().get($self.pos - 1);
        if let Some(token) = token {
            if matches!(token.kind, $kind) {
                Some(token)
            } else {
                $self.pos -= 1;
                None
            }
        } else {
            $self.pos -= 1;
            None
        }
    }};
}

/// Macro to reduce the boilerplate associated with saving the token span of the parsed node.
macro_rules! parse_node {
    ($self:ident, $code:block) => {{
        let start = $self.cur_id();
        let node = { $code };
        $self.ast_ctx.push_span(node.id, start, $self.prev_id());
        Ok(node)
    }};
}

struct Parser {
    input: TokenizedInput,
    ast_ctx: AstContext,

    // state
    pos: usize,
}

impl Parser {
    fn new(input: TokenizedInput) -> Self {
        Self {
            input,
            ast_ctx: AstContext::default(),
            pos: 0,
        }
    }

    #[inline]
    fn tokens(&self) -> &[Rc<Token>] {
        &self.input.tokens()
    }

    fn peek(&self) -> Option<&TokenKind> {
        self.tokens().get(self.pos).map(|t| &t.kind)
    }

    fn next(&mut self) -> Option<Rc<Token>> {
        self.pos += 1;
        let token = self.tokens().get(self.pos - 1).map(|t| t.clone());
        token
    }

    fn cur_id(&self) -> Option<TokenId> {
        self.tokens().get(self.pos).map(|t| t.id)
    }

    fn prev_id(&self) -> Option<TokenId> {
        self.tokens().get(self.pos - 1).map(|t| t.id)
    }

    fn run(&mut self) -> Result<()> {
        while self.pos < self.tokens().len() {
            let start = self.cur_id();
            let stmt = self.parse_stmt()?;

            self.ast_ctx.push_span(stmt.id, start, self.prev_id());
            self.ast_ctx.push_stmt(stmt);
        }

        Ok(())
    }

    fn parse_stmt(&mut self) -> Result<Stmt> {
        let start_pos = self.pos;
        match self.peek() {
            Some(TokenKind::Print) => self.parse_print_stmt(),
            Some(TokenKind::Name(_)) => {
                // here an expr stmt could be confused with an assignment stmt
                let res = self.parse_assign_stmt();
                if res.is_ok() {
                    res
                } else {
                    self.pos = start_pos;
                    self.parse_expr_stmt()
                }
            }
            Some(_) => self.parse_expr_stmt(),
            None => Err(self.end_of_file_err()),
        }
    }

    fn parse_print_stmt(&mut self) -> Result<Stmt> {
        parse_node!(self, {
            expect!(self, TokenKind::Print)?;
            let expr = self.parse_expr()?;
            expect!(self, TokenKind::Semi)?;

            Stmt::print(expr)
        })
    }

    fn parse_assign_stmt(&mut self) -> Result<Stmt> {
        parse_node!(self, {
            let name = self.parse_var_name()?;
            expect!(self, TokenKind::Assign)?;
            let expr = self.parse_expr()?;
            expect!(self, TokenKind::Semi)?;

            Stmt::assign(name, expr)
        })
    }

    fn parse_expr_stmt(&mut self) -> Result<Stmt> {
        parse_node!(self, {
            let expr = self.parse_expr()?;
            expect!(self, TokenKind::Semi)?;

            Stmt::expr(expr)
        })
    }

    fn parse_expr(&mut self) -> Result<Expr> {
        self.parse_add_term()
    }

    fn parse_add_term(&mut self) -> Result<Expr> {
        let mut start = self.cur_id();
        let mut lhs = self.parse_mul_term()?;
        while let Some(op_token) = accept!(self, TokenKind::Plus | TokenKind::Minus) {
            // create binop node and push span
            let op = BinOp::new(&op_token.kind);
            self.ast_ctx
                .push_span(op.id, Some(op_token.id), Some(op_token.id));

            let rhs = self.parse_mul_term()?;
            lhs = Expr::bin_op(op, lhs, rhs);
            self.ast_ctx.push_span(lhs.id, start, self.prev_id());
            start = self.cur_id();
        }
        Ok(lhs)
    }

    fn parse_mul_term(&mut self) -> Result<Expr> {
        let mut start = self.cur_id();
        let mut lhs = self.parse_factor()?;
        while let Some(op_token) = accept!(self, TokenKind::Slash | TokenKind::Star) {
            // create binop node and push span
            let op = BinOp::new(&op_token.kind);
            self.ast_ctx
                .push_span(op.id, Some(op_token.id), Some(op_token.id));

            let rhs = self.parse_factor()?;
            lhs = Expr::bin_op(op, lhs, rhs);
            self.ast_ctx.push_span(lhs.id, start, self.prev_id());
            start = self.cur_id();
        }
        Ok(lhs)
    }

    fn parse_factor(&mut self) -> Result<Expr> {
        match self.peek() {
            Some(TokenKind::Int(_)) => self.parse_integer(),
            Some(TokenKind::Float(_)) => self.parse_float(),
            Some(TokenKind::Name(_)) => self.parse_variable(),
            Some(TokenKind::Minus) | Some(TokenKind::Plus) => self.parse_unary_factor(),
            Some(TokenKind::LParen) => self.parse_grouping(),
            Some(_) => unimplemented!(), //TODO: error handling
            None => Err(self.end_of_file_err()),
        }
    }

    fn parse_integer(&mut self) -> Result<Expr> {
        parse_node!(self, {
            let tok = expect!(self, TokenKind::Int(_))?;
            if let TokenKind::Int(ref n) = tok.kind {
                Expr::integer(n.clone())
            } else {
                unreachable!()
            }
        })
    }

    fn parse_float(&mut self) -> Result<Expr> {
        parse_node!(self, {
            let tok = expect!(self, TokenKind::Float(_))?;
            if let TokenKind::Float(ref n) = tok.kind {
                Expr::float(n.clone())
            } else {
                unreachable!()
            }
        })
    }

    fn parse_variable(&mut self) -> Result<Expr> {
        parse_node!(self, {
            let name = self.parse_var_name()?;
            Expr::variable(name)
        })
    }

    fn parse_unary_factor(&mut self) -> Result<Expr> {
        let start = self.cur_id();
        let op_token = expect!(self, TokenKind::Plus | TokenKind::Minus)?;

        // create binop node
        let op = UnaryOp::new(&op_token.kind);
        self.ast_ctx
            .push_span(op.id, Some(op_token.id), Some(op_token.id));

        let sub_expr = self.parse_factor()?;
        let expr = Expr::unary_op(op, sub_expr);
        self.ast_ctx.push_span(expr.id, start, self.prev_id());

        Ok(expr)
    }

    fn parse_grouping(&mut self) -> Result<Expr> {
        expect!(self, TokenKind::LParen)?;
        let expr = self.parse_expr()?;
        expect!(self, TokenKind::RParen)?;
        Ok(expr)
    }

    fn parse_var_name(&mut self) -> Result<VarName> {
        parse_node!(self, {
            let tok = expect!(self, TokenKind::Name(_))?;
            if let TokenKind::Name(ref n) = tok.kind {
                VarName::new(n.clone())
            } else {
                unreachable!()
            }
        })
    }

    /// Create an EOF error.
    ///
    /// For this, we create a span based on the end location of the last token.
    fn end_of_file_err(&self) -> ParserError {
        let token_id = self.tokens().last().map(|t| t.id);

        let span = if let Some(token_id) = token_id {
            self.input.token_ctx.span_from_id(token_id)
        } else {
            None
        };

        ParserError::SyntaxError(
            SyntaxError::UnexpectedEndOfFile,
            self.input.text_ctx.error_context(span),
        )
    }
}
