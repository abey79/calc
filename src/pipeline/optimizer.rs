use crate::context::checked_ast::{CheckedExpr, CheckedStmt};
use crate::data::ast::{BinOp, BinOpKind, Expr, ExprKind, Stmt, StmtKind};
use crate::errors::OptimizerError;
use crate::states::CheckedState;
use std::ops::{Add, Div, Mul, Sub};

type Result<T> = std::result::Result<T, OptimizerError>;

pub(crate) fn optimize(input: CheckedState) -> CheckedState {
    let optimizer = Optimizer::new(input);
    optimizer.run()
}

struct Optimizer {
    input: CheckedState,
}

// Note:
// As it stands, this object is useless as not local state is needed, and could be replaced by a
// functions. Clippy rightly complains about this, thus the #[allow(only_used_in_recursion)].
// However, improved optimisation would require state (e.g. variable substitution).

#[allow(clippy::only_used_in_recursion)]
impl Optimizer {
    fn new(input: CheckedState) -> Self {
        Self { input }
    }

    fn run(mut self) -> CheckedState {
        let old_stmts: Vec<_> = self.input.ast.stmts_mut().drain(..).collect();
        old_stmts.into_iter().for_each(|stmt| {
            let new_stmt = self.optimize_stmt(stmt);
            self.input.ast.push_stmt(new_stmt);
        });

        self.input
    }

    fn optimize_stmt(&mut self, stmt: CheckedStmt) -> CheckedStmt {
        match stmt.kind {
            StmtKind::Expr { expr } => Stmt::expr(self.optimize_expr(expr), stmt.meta),
            StmtKind::Assign { name, value } => {
                Stmt::assign(name, self.optimize_expr(value), stmt.meta)
            }
            StmtKind::Print { expr } => Stmt::print(self.optimize_expr(expr), stmt.meta),
        }
    }

    fn optimize_expr(&mut self, expr: CheckedExpr) -> CheckedExpr {
        use ExprKind::*;

        match expr.kind {
            BinOp { op, left, right } => {
                let new_left = self.optimize_expr(*left);
                let new_right = self.optimize_expr(*right);

                match (&new_left.kind, &new_right.kind) {
                    (Integer(a), Integer(b)) => Expr::integer(op.eval(*a, *b), expr.meta),
                    (Float(a), Float(b)) => Expr::float(op.eval(*a, *b), expr.meta),
                    _ => Expr::bin_op(op, new_left, new_right, expr.meta),
                }
            }
            _ => expr,
        }
    }
}

impl<M> BinOp<M> {
    fn eval<T>(&self, a: T, b: T) -> T
    where
        T: Add<Output = T> + Sub<Output = T> + Mul<Output = T> + Div<Output = T>,
    {
        match self.kind {
            BinOpKind::Add => a + b,
            BinOpKind::Sub => a - b,
            BinOpKind::Mul => a * b,
            BinOpKind::Div => a / b,
        }
    }
}

#[cfg(test)]
mod test {
    use crate::pipeline::checker::check;
    use crate::pipeline::optimizer::optimize;
    use crate::pipeline::parser::parse;
    use crate::pipeline::tokenizer::tokenize;
    use crate::states::InputState;

    #[test]
    fn test_checker() {
        let input = InputState::from("print (1 + 4) * 3 / (3 + 2);");
        let tokenized = tokenize(input).unwrap();
        let parsed = parse(tokenized).unwrap();
        let checked = check(parsed).unwrap();
        let optimized = optimize(checked);

        insta::assert_debug_snapshot!(optimized.ast);
    }
}
