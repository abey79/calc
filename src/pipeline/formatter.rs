//! Formatter
//!
//! This pipeline stage operates on a generic `Ast<T>` context data, as it doesn't require any AST
//! metadata. This means that it can be run on either `ParsedState` or `CheckedState`.

use crate::context::ast::Ast;
use crate::data::ast::{Expr, ExprKind, Stmt, StmtKind};
use std::fmt;
use std::fmt::{Debug, Display, Write};

pub(crate) fn format<T: Debug + Display, W: Write>(
    input: &Ast<T>,
    writer: &mut W,
) -> Result<(), fmt::Error> {
    let mut formatter = Formatter::new(input, writer);
    formatter.format()?;
    Ok(())
}

struct Formatter<'a, T: Debug + Display, W: Write> {
    input: &'a Ast<T>,
    writer: &'a mut W,
    // here there would be additional state, e.g. indentation level
    // nothing because indentation is not needed for this toy language
}

impl<'a, T: Debug + Display, W: Write> Formatter<'a, T, W> {
    fn new(input: &'a Ast<T>, writer: &'a mut W) -> Self {
        Self { input, writer }
    }

    fn format(&mut self) -> fmt::Result {
        for stmt in self.input.stmts() {
            self.format_stmt(stmt)?;
            writeln!(self.writer)?;
        }

        Ok(())
    }

    fn format_stmt(&mut self, stmt: &Stmt<T>) -> fmt::Result {
        match &stmt.kind {
            StmtKind::Assign { name, value } => {
                write!(self.writer, "{} = ", name)?;
                self.format_expr(value)?;
                write!(self.writer, ";")?;
            }
            StmtKind::Print { expr } => {
                write!(self.writer, "print ")?;
                self.format_expr(expr)?;
                write!(self.writer, ";")?;
            }
            StmtKind::Expr { expr } => {
                self.format_expr(expr)?;
                write!(self.writer, ";")?;
            }
        }

        Ok(())
    }

    fn format_expr(&mut self, expr: &Expr<T>) -> fmt::Result {
        match &expr.kind {
            ExprKind::Variable(name) => write!(self.writer, "{}", name)?,
            ExprKind::BinOp { op, left, right } => {
                let left_precedence = left.kind.precedence();
                let right_precedence = right.kind.precedence();
                let op_precedence = op.kind.precedence();
                if left_precedence < op_precedence {
                    self.format_expr_paren(left)?;
                } else {
                    self.format_expr(left)?;
                }
                write!(self.writer, " {} ", op)?;
                if right_precedence < op_precedence {
                    self.format_expr_paren(right)?;
                } else {
                    self.format_expr(right)?;
                }
            }
            ExprKind::UnaryOp { op, operand } => {
                let op_precedence = op.kind.precedence();
                let operand_precedence = operand.kind.precedence();
                write!(self.writer, "{}", op)?;

                if operand_precedence < op_precedence {
                    self.format_expr_paren(operand)?;
                } else {
                    self.format_expr(operand)?;
                }
            }
            ExprKind::Integer(i) => write!(self.writer, "{}", i)?,
            ExprKind::Float(f) => write!(self.writer, "{:?}", f)?,
        }

        Ok(())
    }

    fn format_expr_paren(&mut self, expr: &Expr<T>) -> fmt::Result {
        write!(self.writer, "(")?;
        self.format_expr(expr)?;
        write!(self.writer, ")")
    }
}

#[cfg(test)]
mod test {
    use crate::pipeline::checker::check;
    use crate::pipeline::formatter::format;
    use crate::pipeline::optimizer::optimize;
    use crate::pipeline::parser::parse;
    use crate::pipeline::tokenizer::tokenize;
    use crate::states::InputState;

    #[test]
    fn test_formatter_from_raw_ast() {
        let input = InputState::from("a = (1.3 + 3.2) * 45.1; b = a * 3.2; print 1 + 2 * 3;");
        let tokenized = tokenize(input).unwrap();
        let parsed = parse(tokenized).unwrap();

        let mut output = String::new();
        format(&parsed.raw_ast, &mut output).unwrap();

        insta::assert_debug_snapshot!(output);
    }

    #[test]
    fn test_formatter_from_checked_ast() {
        let input = InputState::from("a = (1.3 + 3.2) * 45.1; b = a * 3.2; print 1 + 2 * 3;");
        let tokenized = tokenize(input).unwrap();
        let parsed = parse(tokenized).unwrap();
        let checked = check(parsed).unwrap();

        let mut output = String::new();
        format(&checked.ast, &mut output).unwrap();

        insta::assert_debug_snapshot!(output);
    }

    #[test]
    fn test_formatter_from_optimized_ast() {
        let input = InputState::from("a = (1.3 + 3.2) * 45.1; b = a * 3.2; print 1 + 2 * 3;");
        let tokenized = tokenize(input).unwrap();
        let parsed = parse(tokenized).unwrap();
        let checked = check(parsed).unwrap();
        let optimized = optimize(checked);

        let mut output = String::new();
        format(&optimized.ast, &mut output).unwrap();

        insta::assert_debug_snapshot!(output);
    }
}
