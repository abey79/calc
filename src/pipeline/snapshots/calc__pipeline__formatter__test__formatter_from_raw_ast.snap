---
source: src/pipeline/formatter.rs
expression: output
---
a = (1.3 + 3.2) * 45.1;
b = a * 3.2;
print 1 + 2 * 3;

